jetstream_pt/fetch_models.py

```diff
@@ -0,0 +1,209 @@
+
+
+def get_sharding_config(repo_id):
+  #TODO
+  return None
+
+
+def construct_env_data_from_model_id(
+    repo_id, 
+    batch_size, 
+    input_length, 
+    output_length, 
+    quantize, 
+  ):
+  tokenizer_path = os.path.join(_hf_dir(repo_id), 'tokenizer.model')
+  if quantize:
+    checkpoint_path = _int_dir(repo_id)
+    checkpoint_format = 'safetensors'
+  else:
+    checkpoint_path = _bf_dir(repo_id)
+    checkpoint_format = 'safetensors'
+
+  sharding_config = get_sharding_config(repo_id)
+  shard_on_batch = sharding_config is None
+
+  max_cache_length = (
+    FLAGS.override_max_cache_length 
+    if FLAGS.override_max_cache_length > 0 
+    else input_length + output_length)
+
+
+  env_data = JetEngineEnvironmentData(
+      tokenizer_path=tokenizer_path,
+      checkpoint_path=checkpoint_path,
+      checkpoint_format=checkpoint_format,
+      batch_size=batch_size,
+      max_decode_length=output_length,
+      max_input_sequence_length=input_length,
+      # TODO
+      quant_config=QuantizationConfig(),
+      cache_sequence_length=max_cache_length,
+      bf16_enable=True,
+      sharding_config_path=sharding_config,
+      shard_on_batch=shard_on_batch,
+  )
+  model_info = model_id_to_class.get(repo_id)
+  env_data.cache_shape = (
+    batch_size, 
+    model_info.num_heads, 
+    max_cache_length, 
+    model_info.head_dim)
+  env_data.num_layers = model_info.num_layers
+  return env_data
+
+def _load_weights(directory):
+  safetensors_files = glob.glob(os.path.join(directory, '*.safetensors'))
+  state_dict = {}
+  for file_path in safetensors_files:
+      with safe_open(file_path, framework="pt") as f:
+          for key in f.keys():
+              state_dict[key] = f.get_tensor(key)
+  # Load the state_dict into the model
+  return state_dict
+
+
+def instantiate_model_from_repo_id(
+  repo_id,
+    batch_size, 
+    input_length, 
+    output_length, 
+    quantize, 
+  ):
+  model_dir = _hf_dir(repo_id)
+  if not os.path.exists(model_dir) or not os.listdir(model_dir):
+    # no weights has been downloaded 
+    hf_download(repo_id, model_dir, FLAGS.hf_token)
+  env_data = construct_env_data_from_model_id(
+      repo_id, 
+      batch_size, 
+      input_length, 
+      output_length, 
+      quantize, 
+  )
+  env = JetEngineEnvironment(env_data)
+
+  model_info = model_id_to_class.get(repo_id)
+  assert model_info is not None
+
+  model = model_info.model_class.from_hf_model_id(repo_id, env)
+  weights = _load_weights(model_dir)
+  return model, weights
+
+  ## QQ do i need to set the weights onto the model?
+
+
+def hf_download(
+  repo_id: str, dest_directory: str,
+  hf_token: Optional[str] = None) -> None:
+  os.makedirs(dest_directory, exist_ok=True)
+  try:
+      snapshot_download(repo_id, local_dir=dest_directory, 
+                        local_dir_use_symlinks=False, token=hf_token,
+                        allow_patterns=["*.safetensors", "*.json", "*.model"])
```

Comment:
suggestion: change "*.safetensors" to "model*.safetensors", because some repos contain "consolidated" safetensors you do not really need (e.g.: mistral).

Reply:
done.